File: app/src/main/java/com/example/messenger/ChatActivity.java
package com.example.messenger;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class ChatActivity extends AppCompatActivity {

    private TextView textViewTitle;
    private View viewStatus;
    private RecyclerView recyclerViewMessages;
    private EditText editTextMessage;
    private ImageView imageViewSendMessage;
    private MessageAdapter adapter;
    private static final String EXTRA_CURRENT_USER_ID = "user_id";
    private static final String EXTRA_OTHER_USER_ID = "user_other_id";
    private String currentUserId;
    private String otherUserId;
    private ChatViewModel viewModel;
    private ChatViewModelFactory factory;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);
        initViews();
        currentUserId = getIntent().getStringExtra(EXTRA_CURRENT_USER_ID);
        otherUserId = getIntent().getStringExtra(EXTRA_OTHER_USER_ID);
        adapter = new MessageAdapter(currentUserId);
        recyclerViewMessages.setAdapter(adapter);
        factory = new ChatViewModelFactory(currentUserId, otherUserId);
        viewModel = new ViewModelProvider(this, factory).get(ChatViewModel.class);
        observeViewModel();
        imageViewSendMessage.setOnClickListener(l -> {
            String text = editTextMessage.getText().toString().trim();
            Message message = new Message(text, currentUserId, otherUserId);
            viewModel.sendMessage(message);
        });
    }

    private void observeViewModel() {
        viewModel.getError().observe(this, error ->
                Toast.makeText(this, error, Toast.LENGTH_SHORT).show());

        viewModel.getMessages().observe(this,messages ->
                adapter.setMessageList(messages));

        viewModel.getMessageSent().observe(this, isBoolean -> {
            if(isBoolean) {
                editTextMessage.setText("");
            }
        });

        viewModel.getUserOther().observe(this, user -> {
            String userInfo = String.format("%s %s", user.getName(), user.getLastName());
            textViewTitle.setText(userInfo);
        });
    }

    public static Intent newIntent(Context context, String currentUserId, String otherUserId) {
        Intent intent = new Intent(context, ChatActivity.class);
        intent.putExtra(EXTRA_CURRENT_USER_ID, currentUserId);
        intent.putExtra(EXTRA_OTHER_USER_ID, otherUserId);
        return intent;
    }

    private void initViews() {
        textViewTitle = findViewById(R.id.textViewTitle);
        viewStatus = findViewById(R.id.viewStatus);
        recyclerViewMessages = findViewById(R.id.recyclerViewMessages);
        editTextMessage = findViewById(R.id.editTextMessage);
        imageViewSendMessage = findViewById(R.id.imageViewSendMessage);
    }
}

File: app/src/main/java/com/example/messenger/ChatActivity.kt
package com.example.messenger

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import com.example.messenger.adapters.MessageAdapter
import com.example.messenger.databinding.ActivityChatBinding
import com.example.messenger.pojo.Message

class ChatActivity: AppCompatActivity() {
    private lateinit var binding: ActivityChatBinding
    private lateinit var viewModel: ChatViewModel
    private lateinit var adapter: MessageAdapter
    private lateinit var factory: ChatViewModelFactory
    private var currentUserId = ""
    private var otherUserId = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityChatBinding.inflate(layoutInflater)
        setContentView(binding.root)
        otherUserId = intent.getStringExtra(EXTRA_OTHER_USER_ID) ?: return
        currentUserId = intent.getStringExtra(EXTRA_CURRENT_USER_ID) ?: return
        adapter = MessageAdapter(currentUserId)
        binding.recyclerViewMessages.adapter = adapter
        factory = ChatViewModelFactory(currentUserId, otherUserId)
        viewModel = ViewModelProvider(this, factory)[ChatViewModel::class.java]
        observeViewModel()
        binding.imageViewSendMessage.setOnClickListener {
            val text = binding.editTextMessage.text.toString().trim()
            val message = Message(text, currentUserId, otherUserId)
            viewModel.sendMessage(message)
        }
    }

    private fun observeViewModel() {
        viewModel.getError.observe(this) {
            Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
        }
        viewModel.getMessages.observe(this) { adapter.messageList = it }
        viewModel.getMessageSent.observe(this) {
            if(it) binding.editTextMessage.setText("")
        }
        viewModel.getUserOther.observe(this) {
            val userInfo = String.format("%s %s", it.name, it.lastname)
            binding.textViewTitle.text = userInfo
            val backgroundId = if(it.online) R.drawable.circle_green else R.drawable.circle_red
            val background = ContextCompat.getDrawable(this, backgroundId)
            binding.viewStatus.background = background
        }
    }

    companion object {
        private const val EXTRA_CURRENT_USER_ID = "user_id"
        private const val EXTRA_OTHER_USER_ID = "user_other_id"

        fun newIntent(context: Context, currentUserId: String, otherUserId: String): Intent {
            val intent = Intent(context, ChatActivity::class.java)
            intent.putExtra(EXTRA_CURRENT_USER_ID, currentUserId)
            intent.putExtra(EXTRA_OTHER_USER_ID, otherUserId)
            return intent
        }
    }

    override fun onPause() {
        super.onPause()
        viewModel.onlineUserStatus(false)
    }

    override fun onResume() {
        super.onResume()
        viewModel.onlineUserStatus(true)
    }
}